cpu/sam0_common/sam0-eth/eth-netdev.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2020 Mesotic SAS
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_sam0_common
+ * @{
+ *
+ * @file
+ * @brief       Low-level Ethernet driver implementation
+ *
+ * @author      Dylan Laduranty <dylan.laduranty@mesotic.com>
+ *
+ * @}
+ */
+#include <string.h>
+
+#include "mutex.h"
+#include "luid.h"
+
+#include "iolist.h"
+#include "net/gnrc/netif/ethernet.h"
+#include "net/gnrc.h"
+#include "net/ethernet.h"
+#include "net/netdev/eth.h"
+
+#include "periph/gpio.h"
+
+#include "sam0_eth_netdev.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+#include "log.h"
+
+/* Internal helpers */
+int sam0_eth_init(void);
+int sam0_eth_send(const struct iolist *iolist);
+int sam0_eth_receive_blocking(char *data, unsigned max_len);
+void sam0_eth_set_mac(const char *mac);
+void sam0_eth_get_mac(char *out);
+
+sam0_eth_netdev_t _sam0_eth_dev;
+
+static int _sam0_eth_init(netdev_t *netdev)
+{
+    (void)netdev;
+    sam0_eth_init();
+    return 0;
+}
+
+static void _sam0_eth_isr(netdev_t *netdev)
+{
+    netdev->event_callback(netdev, NETDEV_EVENT_RX_COMPLETE);
+    return;
+}
+
+static int _sam0_eth_recv(netdev_t *netdev, void *buf, size_t len, void *info)
+{
+    (void)info;
+    (void)netdev;
+    mutex_lock(&_sam0_eth_dev.dev_lock);
+    unsigned ret = sam0_eth_receive_blocking((char *)buf, len);
+    mutex_unlock(&_sam0_eth_dev.dev_lock);
+
+    return ret;
+}
+
+static int _sam0_eth_send(netdev_t *netdev, const iolist_t *iolist)
+{
+    int ret;
+
+    mutex_lock(&_sam0_eth_dev.dev_lock);
+
+    netdev->event_callback(netdev, NETDEV_EVENT_TX_STARTED);
+    ret = sam0_eth_send(iolist);
+    if (ret == -EOVERFLOW) {
+        mutex_unlock(&_sam0_eth_dev.dev_lock);
+        /* TODO: use a specific netdev callback here ? */
+        return -EOVERFLOW;
```

Comment:
```suggestion
        return -EBUSY;
```

If `gnrc_netif_pktq` this will re-try the send when TX DONE is reported.